src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6845,75 +6845,66 @@ void reportBadDelegateParameter(BindingDiagnosticBag bag, MethodSymbol sourceInv
             }
         }
 
-        private void ReportNullabilityMismatchWithTargetDelegate(Location location, NamedTypeSymbol delegateType, UnboundLambda unboundLambda)
+        private void ReportNullabilityMismatchWithTargetDelegate(Location location, NamedTypeSymbol delegateType, BoundLambda lambda)
         {
-            if (!unboundLambda.HasExplicitlyTypedParameterList)
-            {
-                return;
-            }
+            MethodSymbol? targetInvokeMethod = delegateType.DelegateInvokeMethod;
+            LambdaSymbol sourceInvokeMethod = lambda.Symbol;
+            UnboundLambda unboundLambda = lambda.UnboundLambda;
 
-            var invoke = delegateType?.DelegateInvokeMethod;
-            if (invoke is null)
+            if (targetInvokeMethod is null ||
+                targetInvokeMethod.ParameterCount != sourceInvokeMethod.ParameterCount)
             {
                 return;
             }
 
-            Debug.Assert(delegateType is object);
-            if (unboundLambda.Syntax is LambdaExpressionSyntax lambdaSyntax)
+            // Parameter nullability is expected to match exactly. This corresponds to the behavior of initial binding.
```

Comment:
It might be good to update this comment to indicate more clearly that we call `CheckValidNullableMethodOverride` "bidirectionally" to simulate an "invariant" check.